### PR TITLE
Implement a validity check for firstSN

### LINKED ISSUE #3236: Assertion failure in SequenceNumber.h via malformed SPDP packet only when compiled in logging-enabled (Debug) mode
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

Malformed submessages handled gracefully.

### Current behavior

We came across two following issues.
1. An SPDP payload with a malformed heartbeat submessage triggers an assertion failure at `fastrtps/include/fastdds/rtps/common/SequenceNumber.h:247`.
2. This only happens when fastrtps is compiled with the following CMake arguments for logging:
```
--cmake-args -DCMAKE_BUILD_TYPE=Debug -DLOG_NO_INFO=OFF -DINTERNAL_DEBUG=ON -DLOG_CONSUMER_DEFAULT=STDOUT -DLOG_NO_WARNING=OFF -DLOG_NO_ERROR=OFF
```

### Steps to reproduce

1. Build fastdds v2.9.0 and HelloWorldExample with logging (Debug mode)
```
mkdir -p ~/fastdds-log/src
cd ~/fastdds-log
wget https://raw.githubusercontent.com/eProsima/Fast-DDS/v2.9.0/fastrtps.repos
vcs import src < fastrtps.repos
colcon build --cmake-args -DCMAKE_BUILD_TYPE=Debug -DLOG_NO_INFO=OFF -DINTERNAL_DEBUG=ON -DLOG_CONSUMER_DEFAULT=STDOUT -DLOG_NO_WARNING=OFF -DLOG_NO_ERROR=OFF
source install/setup.sh
cd src/fastrtps/examples/cpp/dds/HelloWorldExample
cmake .
make
```

2. Build fastdds v2.9.0 and HelloWorldExample without logging
```
mkdir -p ~/fastdds-nolog/src
cd ~/fastdds-nolog
wget https://raw.githubusercontent.com/eProsima/Fast-DDS/v2.9.0/fastrtps.repos
vcs import src < fastrtps.repos
colcon build
source install/setup.sh
cd src/fastrtps/examples/cpp/dds/HelloWorldExample
cmake .
make
```

3. Launch `DDSHelloWorldExample` with logging and send the malformed packet

    * Launching:
    ```
    source ~/fastdds-log/install/setup.sh && ~/fastdds-log/src/fastrtps/examples/cpp/dds/HelloWorldExample/DDSHelloWorldExample publisher
    ```

    * Hexdump of the SPDP packet to send (to `239.255.0.1:7400`):
    ```
    0000  45 00 00 68 00 01 40 00 40 11 D9 76 80 3D F0 CF  E..h..@.@..v.=..
    0010  EF FF 00 01 05 39 1C E8 00 54 08 B9 52 54 50 53  .....9...T..RTPS
    0020  02 04 01 0F 01 03 02 42 AC 11 00 02 45 E5 E2 FD  .......B....E...
    0030  07 BF 34 00 00 00 10 00 00 00 00 00 00 01 00 C2  ..4.............
    0040  00 00 00 00 A7 9B EA 8C BE ED DC CB 00 03 00 00  ................
    0050  77 00 04 00 00 00 00 00 00 00 00 00 00 00 00 00  w...............
    0060  00 00 00 00 00 00 00 00                          ........
    ```

    * Result after sending the packet:
    ```
    Starting 
    Publisher running 10 samples.
    DDSHelloWorldExample: /home/seulbae/fastdds-log/src/fastrtps/include/fastdds/rtps/common/SequenceNumber.h:247: eprosima::fastrtps::rtps::SequenceNumber_t eprosima::fastrtps::rtps::operator-(const eprosima::fastrtps::rtps::SequenceNumber_t&, uint32_t): Assertion `0 < res.high' failed.
    [1]    3561127 abort       publisher
    ```

    Assertion failure has been triggered!

4. Launch `DDSHelloWorldExample` without logging and send the malformed packet

    * Launching:
    ```
    source ~/fastdds-nolog/install/setup.sh && ~/fastdds-nolog/src/fastrtps/examples/cpp/dds/HelloWorldExample/DDSHelloWorldExample publisher
    ```

    * Send the same SPDP packet

    * Result after sending:
    ```
    Starting 
    Publisher running 10 samples.
    ```

    No assertion failure this time..

### Fast DDS version/commit

Tag v2.9.0 (4c55488)

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

Alternatively, you can directly use [this PoC](https://gist.github.com/squizz617/e558b28c1005c6ad96db1856f49d8263) to quickly test the behavior.

But make sure to compile this with an afl compiler (e.g., `afl-clang-fast++`) or manually enable the `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` flag to keep `participantGuidPrefix` uninitialized.

Thank you.

### XML configuration file

_No response_

### Relevant log output

_No response_

### Network traffic capture

_No response_

### PR DESCRIPTION
**Pull request by squizz617**
This commit suggests a fix for issue #3236.

## Description

As per 8.3.8.6.3 of DDS-RTPS 2.5, "Heartbeat submessage is invalid if firstSN.value is zero or negative". However, a check for this property is missing. This leads to an assertion failure when an invalid firstSN is provided (see issue #3236).

@Mergifyio backport 2.9.x

Fixes #3236 

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally -> 
- (N/A) Any new/modified methods have been properly documented using Doxygen. 
- (N/A) Changes are ABI compatible. 
- (N/A) Changes are API compatible. 
- (N/A) Documentation builds and tests pass locally. 
- (N/A) New feature has been added to the `versions.md` file (if applicable).
- (N/A) New feature has been documented/Current behavior is correctly described in the documentation. 

- [x] Applicable backports have been included in the description.

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/messages/MessageReceiver.cpp [hunk 1]
```diff
@@ -1056,6 +1056,13 @@ bool MessageReceiver::proc_Submsg_Heartbeat(
     SequenceNumber_t lastSN;
     CDRMessage::readSequenceNumber(msg, &firstSN);
     CDRMessage::readSequenceNumber(msg, &lastSN);
+
+    SequenceNumber_t zeroSN;
+    if (fisrtSN <= zeroSN)
```
Comment (MiguelCompany) [inline src/cpp/rtps/messages/MessageReceiver.cpp:1061]: This is not building
```suggestion
    if (firstSN <= zeroSN)
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
@squizz617 Thank you for your contribution!

This is not building (see my comment below).

**Comment by squizz617** [comment]
Sorry, made a mistake while moving stuffs into the pr branch :(
Fixed and pushed!

**Comment by squizz617** [comment]
Hi @MiguelCompany, I'm just wondering if you've looked at the changes. Thank you.

**Comment by MiguelCompany** [comment]
> Hi @MiguelCompany, I'm just wondering if you've looked at the changes. Thank you.

@squizz617 Changes look good to me. Is it possible that you could add a regression test for this?

I think that adding a new file [here](https://github.com/eProsima/Fast-DDS/tree/master/fuzz/C%2B%2B/fuzz_processCDRMsg/fuzz_processCDRMsg_seed_corpus) would be enough

**Comment by squizz617** [comment]
@MiguelCompany Done! I temporarily named the file `input_issue3236`, but please change if needed.

**Review by MiguelCompany — APPROVED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@mergifyio backport 2.9.x 2.8.x 2.6.x

**Pull request merged**